README.md
aurora v0.0.3
============ 

Control ws2812b pixel with an atmege328p.
The tcp stack based on the tuxgraphic tcp stack (http://tuxgraphics.org/electronics/).

You can change the color of the pixel with an simple tcp message, like a hex string.
Use an individual tcp-server or node-red.
Also see my other project named 'sphere'. This is a nodejs based http-server to 
control several aurora server.

Externals
============ 
 * light_ws2812 V2.4 (light_ws2812_AVR)
 * eth_lcd_tcp-2.10

Hardware
================
 
 * atmega328p nano board
 * enc28J60 nano Ethernet shield
 * ws2812b 300pxl
 * lcd - 2x16 hd44780
 
wiring
---------------------

| __**atmega328p**__         | __**enc28J60**__      | __**lcd 2x16**__          | __**ws2812b**__      |
|----------------------------|-----------------------|---------------------------|----------------------|
|                            |                       |                           |                      |
| PD2                        |                       | PIN4  - RS (v0.0.1 only)  |                      |
| PD3                        |                       | PIN6  - E  (v0.0.1 only)  |                      |
| PD4                        |                       | PIN11 - D4 (v0.0.1 only)  |                      |
| PD5                        |                       | PIN12 - D5 (v0.0.1 only)  |                      |
| PD6                        |                       | PIN13 - D6 (v0.0.1 only)	 |                      |
| PD7                        |                       | PIN14 - D7 (v0.0.1 only)	 |                      |
| PD8                        |                       |                           | DIN                  |
|                            |                       |                           |                      |
| PD10                       | SS                    |                           |                      |
| PD11                       | MOSI                  |                           |                      |
| PD12                       | MISO                  |                           |                      |
| PD12                       | SCK                   |                           |                      |
|                            |                       |                           |                      |



Getting started
================

 You can use the make file after git clone.
 The default ipAddress is 10.11.0.100, change this in aurora.c
 `static uint8_t myip[4] = {10,11,0,100};      //10.11.0.100`

 gcc version 4.8.2 (GCC) 

API description:
---------------------

	
# standard request:
| __**byte**__  | __**description**__   |
|---------------|-----------------------|
| 0-1           | 4	(low byte first)	  |
| 2             | CMD								    |
| 3             | port								  |


# error request:

| __**byte**__  | __**description**__   |
|---------------|-----------------------|
| 0-1           | 4	(low byte first)	  |
| 2             | 0xff								  |
| 3             | error code					  |


# commands

# clearPixel:

clear the internal buffer of led matrix

| __**byte**__  | __**description**__ 	    |
|---------------|---------------------------|
| 0-1           | 3	length (low byte first) |
| 2             | CMD_CLEARPIXEL = 0x12	    |


# setColor:

set the color of all pixel. 
Pixels hardcoded as define 'MAX_PIXEL'.

| __**byte**__  | __**description**__ 	    |
|---------------|---------------------------|
| 0-1           | 7	length (low byte first) |
| 2             | CMD_SETCOLOR = 0x15	      |
| 3             | PIN									      |
| 4             | color value red 		      |
| 5             | color value green 	      |
| 6             | color value blue		      |

# setPixel:

set the color of an specific pixel.
 
| __**byte**__  | __**description**__   |
|---------------|-----------------------|
| 0-1           | 9	(low byte first)	  |
| 2             | CMD_SETPIXEL = 0x13	  |
| 3             | PIN									  |
| 4-5           | pixel (low byte first)|
| 6             | color value red 		  |
| 7             | color value green 	  |
| 8             | color value blue		  |

# setRange:

set the color of an specific range.
 
| __**byte**__  | __**description**__   |
|---------------|-----------------------|
| 0-1           | 10	(low byte first)  |
| 2             | CMD_SETRANGE = 0x14	  |
| 3             | PIN									  |
| 4-5           | first pixel (LBGF)    |
| 6-7           | last pixel (LBGF)     |
| 8             | color value red 		  |
| 9             | color value green 	  |
| 10            | color value blue		  |

controller
---------------------

 * [sphere](https://github.com/zombinary/sphere), browser based gui
 * [gravity-aurora](https://github.com/zombinary/gravity-aurora), npm package
 

# Supported Hardware


| __**mcu**__                | __**controller**__    |__**versiont**__           |                           |
|----------------------------|-----------------------|---------------------------|---------------------------|
|                    	     |                       |                           |                           |
| atmega328p                 | ws2812b (300pxl)      |  v0.0.1                   |  :white_check_mark:       |
|                            |      		     |  v0.0.2                   |  :white_check_mark:       |

Road-map
=========
  
  * * add digital input/output




RELEASE_NOTES.md
version 0.0.3
============
  - add clearPixel()/setClearPixel_req();
  - add setPixel()/setPixel_req();
  - add setColor()/setColor_req();
	- add setError_req();
	
version 0.0.2
============
  - change color value format hexstring ==> hex
  - remove lcd output

version 0.0.1
============

  - receive color value as hexstring and
    push it to led stripe with fix pixel length

aurora.c
/*********************************************
 * aurora.c
 * author: zombinary
 * copyright: GPL V2
 * see http://www.gnu.org/licenses/gpl.html
 *
 * tcp client, to control ws2812b pixel.
 *
 * http://tuxgraphics.org/electronics/
 * Note: there is a version number in the text. Search for "tuxgraphics"
 *********************************************/
#include <avr/io.h>

#include <stdint.h>
#include <string.h>
#include <stdlib.h>
#include <avr/interrupt.h>

#include "eth_lcd_tcp-2.10/enc28j60.h"
#include "eth_lcd_tcp-2.10/ip_arp_udp_tcp.h"
//#include "eth_lcd_tcp-2.10/lcd.h"
#include "eth_lcd_tcp-2.10/net.h"
#include "eth_lcd_tcp-2.10/timeout.h"

#include "ws2812b/light_ws2812.h"
//#include "wrapper/wrapper.h"
#include "lib/pixel.h"

#define MAX_PIXEL 15

//tcp settings
static uint8_t mymac[6] = {0x54,0x55,0x58,0x10,0x00,0x65};
static uint8_t myip[4] = {10,11,0,101};
static uint16_t mywwwport =80; // listen port for tcp/www (max range 1-254)

#define BUFFER_SIZE 750
static uint8_t buf[BUFFER_SIZE+1];
#define STR_BUFFER_SIZE 22

struct cRGB led[MAX_PIXEL];   //led[0].r=16;led[0].g=16;led[0].b=16;  white

int main(void){
  uint16_t plen;
  uint8_t dat_p;
   char *request;
   uint8_t port = 0;
   uint16_t pixelpos = 0;
  uint16_t pxl=0;
  uint16_t i = 0;
  uint16_t l = 0;
  uint8_t cmd = 0;

  uint16_t x = 0;
  uint8_t pos = 0;


  initPixel(MAX_PIXEL);
  // init strip - create an array with length of ledstrip (pixel)
  for(x=MAX_PIXEL; x>0; x--)
  {
      led[x-1].r=0;led[x-1].g=0;led[x-1].b=0;
  }

   	/*initialize enc28j60*/
    enc28j60Init(mymac);
    enc28j60clkout(2); // change clkout from 6.25MHz to 12.5MHz
    _delay_loop_1(50); // 12ms
    enc28j60PhyWrite(PHLCON,0x476);
    _delay_loop_1(50); // 12ms

    /* lcd display: */
    //lcd_init(LCD_DISP_ON);
    //lcd_clrscr();
    //lcd_puts_P("aurora");
    init_ip_arp_udp_tcp(mymac,myip,mywwwport);

  while(1){
        // get the next new packet:
         plen = enc28j60PacketReceive(BUFFER_SIZE, buf);

         // plen will ne unequal to zero if there is a valid
         // packet (without crc error)
         if(plen==0){
           continue;
         }

         // arp is broadcast if unknown but a host may also
         // verify the mac address by sending it to
         // a unicast address.
         if(eth_type_is_arp_and_my_ip(buf,plen)){
             make_arp_answer_from_request(buf);
             continue;
         }

         // check if ip packets are for us:
         if(eth_type_is_ip_and_my_ip(buf,plen)==0){
             continue;
         }

         if(buf[IP_PROTO_P]==IP_PROTO_ICMP_V && buf[ICMP_TYPE_P]==ICMP_TYPE_ECHOREQUEST_V){
             // a ping packet, let's send pong
             make_echo_reply_from_request(buf,plen);
             continue;
         }

         // tcp port start, compare only the lower byte
         if (buf[IP_PROTO_P]==IP_PROTO_TCP_V&&buf[TCP_DST_PORT_H_P]==0&&buf[TCP_DST_PORT_L_P]==mywwwport){
                 if (buf[TCP_FLAGS_P] & TCP_FLAGS_SYN_V){
                         make_tcp_synack_from_syn(buf);
                         // make_tcp_synack_from_syn does already send the syn,ack
                         continue;
                 }
                 if (buf[TCP_FLAGS_P] & TCP_FLAGS_ACK_V){
                         init_len_info(buf); // init some data structures
                         // we can possibly have no data, just ack:
                         dat_p=get_tcp_data_pointer();
                         if (dat_p==0){
                          if (buf[TCP_FLAGS_P] & TCP_FLAGS_FIN_V){
                            // finack, answer with ack
                                  make_tcp_ack_from_any(buf);
                             }
                             // just an ack with no data, wait for next packet
                          continue;
                         }
                         if(buf[TCP_FLAGS_P] & TCP_FLAGS_PSHACK_V){
                          if (buf[TCP_FLAGS_P] & TCP_FLAGS_ACK_V){
                             init_len_info(buf); // init some data structures
                             	 	 	 	 	 // we can possibly have no data, just ack:
                             dat_p=get_tcp_data_pointer();
                             if (dat_p==0){
                                if (buf[TCP_FLAGS_P] & TCP_FLAGS_FIN_V){
                                   // finack, answer with ack
                                    make_tcp_ack_from_any(buf);
                                }
                                 // just an ack with no data, wait for next packet
                                 continue;
                             }

                             if (strncmp("GET ",(char *)&(buf[dat_p]),4) == 0){
                            	 plen=fill_tcp_data_p(buf,0,PSTR("HTTP/1.0 200 OK\r\nContent-Type: text/html\r\n\r\n<h1>aurora	</h1>"));
                            	 make_tcp_ack_from_any(buf); // send ack for http get
                                 make_tcp_ack_with_data(buf,plen); // send data

                             }else{
                            	 cmd = buf[dat_p+2];
                            	 switch(cmd){
                            	 	 case 0x12:
                            	 		  //CMD_CLEAR
                            	 		  clearPixel((uint8_t *)&led);
                            	          plen=setClearPixel_req(buf);
                            	          break;
                            	 	 case 0x13:
                            	 		 //CMD_SETPIXEL
                            		 	 setPixel((uint8_t *)&led, (uint8_t *)&buf[dat_p]);
                            		 	 plen=setPixel_req(buf);
                            	 		 break;
                            	 	case 0x14:
                            	 		//CMD_SETRANGE
                            	 		setRange((uint8_t *)&led, (uint8_t *)&buf[dat_p]);
                            	 		plen=setRange_req(buf);
                            	 		break;
	                        	 	 case 0x15:
	                        	 		 //CMD_SETCOLOR
                            	 		 setColor((uint8_t *)&led, (uint8_t *)&buf[dat_p]);
                            	 		plen=setColor_req(buf);
                            	 		break;
                            	 	 default:
                            	 		 plen=setError_req(buf, 0x01);
                            	 		 break;

                            	 }
                           		 make_tcp_ack_from_any(buf);
                            	 make_tcp_ack_with_data(buf,plen); // send data
                            	 continue;
                            }
                             make_tcp_ack_with_data(buf,plen); // send data
                             continue;
                          }
                        }
                        continue;
                 }
         }
  }
  return 0;
}
